### ISSUE TITLE
Pass source opts at initialization

### ISSUE THREAD

**Issue by cbarrete** [hrsh7th/nvim-cmp#203]
`sources[number].opts` is nice but as far as I can tell, it is only passed to the source for callbacks such as `complete`.

I'd like to pass some configuration to my source when it is first constructed, so that it can initialize itself properly (in my case, I need a directory name in which the files need to be parsed to extract the completion information).

Is there already a way to do this that I'm missing? If not, would you be open to adding one?
I suppose an obvious solution would be to pass `opts` to the source's `new` function, if defined.

**Comment by hrsh7th** [comment]
It's difficult a bit.

1. The source registration can be lazy.
2. We can't detect when to call `cmp.setup`

So it can't realize as deterministic.

**Comment by hrsh7th** [comment]
If you have any idea. I hope to know it.

**Comment by windwp** [comment]
you can use the old way by set a value on vim.g  -------

**Comment by cbarrete** [comment]
>you can use the old way by set a value on vim.g
Indeed, I haven't had a better idea. I've seen some other people set a 
variable in a `config` module, but that's more involved, which I don't 
prefer.

Thanks!
